File: app/src/main/java/com/example/bunnix/database/firebase/collections/ProductCollection.kt
package com.example.bunnix.database.firebase.collections

import com.example.bunnix.database.config.FirebaseConfig
import com.example.bunnix.database.models.Product
import com.google.firebase.firestore.Query
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.tasks.await

object ProductCollection {

    private val collection = FirebaseConfig.firestore.collection(FirebaseConfig.Collections.PRODUCTS)

    // GET ALL PRODUCTS (Real-time)
    fun getAllProducts(): Flow<List<Product>> = callbackFlow {
        val listener = collection
            .whereEqualTo("inStock", true)
            .orderBy("createdAt", Query.Direction.DESCENDING)
            .addSnapshotListener { snapshot, error ->
                if (error != null) {
                    close(error)
                    return@addSnapshotListener
                }
                val products = snapshot?.toObjects(Product::class.java) ?: emptyList()
                trySend(products)
            }
        awaitClose { listener.remove() }
    }

    // GET PRODUCTS BY VENDOR
    fun getProductsByVendor(vendorId: String): Flow<List<Product>> = callbackFlow {
        val listener = collection
            .whereEqualTo("vendorId", vendorId)
            .orderBy("createdAt", Query.Direction.DESCENDING)
            .addSnapshotListener { snapshot, error ->
                if (error != null) {
                    close(error)
                    return@addSnapshotListener
                }
                val products = snapshot?.toObjects(Product::class.java) ?: emptyList()
                trySend(products)
            }
        awaitClose { listener.remove() }
    }

    // GET PRODUCTS BY CATEGORY
    fun getProductsByCategory(category: String): Flow<List<Product>> = callbackFlow {
        val listener = collection
            .whereEqualTo("category", category)
            .whereEqualTo("inStock", true)
            .addSnapshotListener { snapshot, error ->
                if (error != null) {
                    close(error)
                    return@addSnapshotListener
                }
                val products = snapshot?.toObjects(Product::class.java) ?: emptyList()
                trySend(products)
            }
        awaitClose { listener.remove() }
    }

    // ADD PRODUCT
    suspend fun addProduct(product: Product): Result<String> {
        return try {
            val docRef = collection.add(product).await()
            Result.success(docRef.id)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    // UPDATE PRODUCT
    suspend fun updateProduct(productId: String, updates: Map<String, Any>): Result<Unit> {
        return try {
            collection.document(productId).update(updates).await()
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    // DELETE PRODUCT
    suspend fun deleteProduct(productId: String): Result<Unit> {
        return try {
            collection.document(productId).delete().await()
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
}

File: src/database/firebase/collections/productCollection.js
import { db, COLLECTIONS } from '../config';
import { 
  collection, 
  doc,
  addDoc, 
  getDoc,
  getDocs, 
  updateDoc,
  deleteDoc,
  query, 
  where, 
  orderBy,
  limit,
  onSnapshot,
  increment,
  serverTimestamp 
} from 'firebase/firestore';

const productsCollection = collection(db, COLLECTIONS.PRODUCTS);

/**
 * ADD PRODUCT
 */
export const addProduct = async (productData) => {
  try {
    const docRef = await addDoc(productsCollection, {
      ...productData,
      createdAt: serverTimestamp(),
      updatedAt: serverTimestamp()
    });
    return { success: true, productId: docRef.id };
  } catch (error) {
    return { success: false, error: error.message };
  }
};

/**
 * GET PRODUCT BY ID
 */
export const getProduct = async (productId) => {
  try {
    const productDoc = await getDoc(doc(productsCollection, productId));
    if (productDoc.exists()) {
      return { success: true, data: { id: productDoc.id, ...productDoc.data() } };
    } else {
      return { success: false, error: 'Product not found' };
    }
  } catch (error) {
    return { success: false, error: error.message };
  }
};

/**
 * GET ALL PRODUCTS (Real-time)
 */
export const subscribeToAllProducts = (callback) => {
  const q = query(
    productsCollection, 
    where('inStock', '==', true),
    where('isActive', '==', true),
    orderBy('createdAt', 'desc')
  );
  
  return onSnapshot(q, (snapshot) => {
    const products = snapshot.docs.map(doc => ({
      id: doc.id,
      ...doc.data()
    }));
    callback(products);
  });
};

/**
 * GET PRODUCTS BY VENDOR (Real-time)
 */
export const subscribeToVendorProducts = (vendorId, callback) => {
  const q = query(
    productsCollection,
    where('vendorId', '==', vendorId),
    orderBy('createdAt', 'desc')
  );
  
  return onSnapshot(q, (snapshot) => {
    const products = snapshot.docs.map(doc => ({
      id: doc.id,
      ...doc.data()
    }));
    callback(products);
  });
};

/**
 * GET PRODUCTS BY CATEGORY (Real-time)
 */
export const subscribeToProductsByCategory = (category, callback) => {
  const q = query(
    productsCollection,
    where('category', '==', category),
    where('inStock', '==', true),
    where('isActive', '==', true),
    orderBy('createdAt', 'desc')
  );
  
  return onSnapshot(q, (snapshot) => {
    const products = snapshot.docs.map(doc => ({
      id: doc.id,
      ...doc.data()
    }));
    callback(products);
  });
};

/**
 * SEARCH PRODUCTS BY NAME
 */
export const searchProducts = async (searchTerm) => {
  try {
    const snapshot = await getDocs(productsCollection);
    const products = snapshot.docs
      .map(doc => ({ id: doc.id, ...doc.data() }))
      .filter(product => 
        product.name.toLowerCase().includes(searchTerm.toLowerCase()) ||
        product.tags.some(tag => tag.toLowerCase().includes(searchTerm.toLowerCase()))
      );
    return { success: true, data: products };
  } catch (error) {
    return { success: false, error: error.message };
  }
};

/**
 * UPDATE PRODUCT
 */
export const updateProduct = async (productId, updates) => {
  try {
    await updateDoc(doc(productsCollection, productId), {
      ...updates,
      updatedAt: serverTimestamp()
    });
    return { success: true };
  } catch (error) {
    return { success: false, error: error.message };
  }
};

/**
 * DELETE PRODUCT
 */
export const deleteProduct = async (productId) => {
  try {
    await deleteDoc(doc(productsCollection, productId));
    return { success: true };
  } catch (error) {
    return { success: false, error: error.message };
  }
};

/**
 * UPDATE STOCK
 */
export const updateStock = async (productId, newStock) => {
  try {
    await updateDoc(doc(productsCollection, productId), {
      totalStock: newStock,
      inStock: newStock > 0,
      updatedAt: serverTimestamp()
    });
    return { success: true };
  } catch (error) {
    return { success: false, error: error.message };
  }
};

/**
 * INCREMENT VIEWS
 */
export const incrementViews = async (productId) => {
  try {
    await updateDoc(doc(productsCollection, productId), {
      views: increment(1)
    });
    return { success: true };
  } catch (error) {
    return { success: false, error: error.message };
  }
};

/**
 * INCREMENT SOLD
 */
export const incrementSold = async (productId, quantity) => {
  try {
    await updateDoc(doc(productsCollection, productId), {
      sold: increment(quantity),
      totalStock: increment(-quantity)
    });
    return { success: true };
  } catch (error) {
    return { success: false, error: error.message };
  }
};